src/sentry/static/sentry/app/utils/withEnvironmentInQueryString.jsx

```diff
@@ -0,0 +1,81 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import createReactClass from 'create-react-class';
+import Reflux from 'reflux';
+import {browserHistory} from 'react-router';
+import qs from 'query-string';
+
+import LatestContextStore from '../stores/latestContextStore';
+
+const withEnvironmentInQueryString = WrappedComponent =>
+  createReactClass({
+    displayName: 'withEnvironmentInQueryString',
+
+    propTypes: {
+      location: PropTypes.object,
+    },
+
+    mixins: [Reflux.listenTo(LatestContextStore, 'onLatestContextChange')],
+
+    getInitialState() {
+      const latestContext = LatestContextStore.getInitialState();
+      return {
+        environment: latestContext.environment,
+        organization: latestContext.organization,
+        hasEnvironmentsFeature: this.hasEnvironmentsFeature(latestContext.organization),
+      };
+    },
+
+    componentWillMount() {
+      if (this.state.hasEnvironmentsFeature) {
+        const {query, pathname} = this.props.location;
+
+        if (this.state.environment) {
+          const envName = this.state.environment.name;
+          const hasValidEnvironmentInQuery =
+            'environment' in query && query.environment === envName;
+
+          // Update environment in browser history if it is not in sync with the currently active one
+          if (!hasValidEnvironmentInQuery) {
+            query.environment = this.state.environment.name;
+            browserHistory.replace(`${pathname}?${qs.stringify(query)}`);
+          }
+        }
+      }
+    },
+
+    onLatestContextChange({environment, organization}) {
```

Comment:
It looks like we only use `organization` here for feature tag check, should we add a TODO to clean this up when this is released for all?

Reply:
added